finagle-core/src/main/scala/com/twitter/finagle/dispatch/ServerDispatcher.scala

```diff
@@ -103,8 +113,29 @@ class SerialServerDispatcher[Req, Rep](
     service.close()
   }
 
-  protected def dispatch(req: Req, eos: Promise[Unit]) =
+  override def dispatch(req: Req, eos: Promise[Unit]) =
     service(req) ensure eos.setDone()
 
-  protected def handle(rep: Rep) = trans.write(rep)
+  override def handle(rep: Rep) = trans.write(rep)
+}
+
+/**
+  * Wraps a
+  * [[com.twitter.finagle.dispatcher.ServerDispatcher ServerDispatcher]]
+  * adding a
+  * [[com.twitter.finagle.tracing.Annotation.WireRecv WireRecv]]
+  * when a request is read from the wire, and later a
+  * [[com.twitter.finagle.tracing.Annotation.WireSend WireSend]]
+  * when it writes successfully to the wire.
+ */
+final class AnnotatedServerDispatcher[Req, Rep, Out]
+  (dispatcher: ServerDispatcher[Req, Rep, Out]) extends ServerDispatcher[Req, Rep, Out] {
+
+  override def handle(rep: Rep): Future[Unit] =
+    dispatcher.handle(rep).onSuccess { _ => Trace.record(Annotation.WireSend) }
```

Comment:
in order to avoid the closure allocation on every call, we can store it in a member variable.

```
private[this] val RecordWireSend: Unit => Unit = 
  _ => Trace.record(Annotation.WireSend) 

...
dispatcher.handle(rep).onSuccess(RecordWireSed)